### PR TITLE
feat: Add ClpIrFileAppender, AbstractBufferedRollingFileAppender, and AbstractClpIrBufferedRollingFileAppender.

### PR DESCRIPTION
**Pull request by p1halani**
# Description

This PR:
- Adds the following appenders for log4j2 with clp compression and pattern layout
  - AbstractBufferedRollingFileAppender
  - AbstractClpIrBufferedRollingFileAppender
  - ClpIrFileAppender

It uses `clp-ffi` and `zstd-jni` for compression.

# Validation performed

- Tested by locally building and printing the logs via custom Java application. Compressed log files were generated.

## Summary by CodeRabbit

- **New Features**
	- Introduced new logging classes to enhance log management and compression.
	- Added a custom Log4j appender for writing logs into Zstandard-compressed CLP IR stream files.
	- Implemented a utility for creating output file streams with directory creation support.
	- Added support for managing log freshness with configurable timeouts.
	- Introduced a new interface for flexible time management in logging scenarios.
	- Added a new XML configuration for excluding specific bugs from static analysis.

- **Bug Fixes**
	- Enhanced error handling in logging operations to ensure robustness.

- **Documentation**
	- Added documentation for new classes and methods to improve usability and understanding of logging functionalities.

- **Chores**
	- Removed placeholder class `Boilerplate.java` as it was no longer needed.

### CODE REVIEW

File: src/main/java/com/yscope/logging/log4j2/Utils.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        } catch (FileNotFoundException ex) {
+            // Create the parent directory if necessary
+            String parentPath = new File(filePath).getParent();
+            if (null == parentPath) { throw ex; }
+            Files.createDirectories(Paths.get(parentPath));
+
+            fileOutputStream = new FileOutputStream(filePath, false);
+        }
+        return fileOutputStream;
+    }
```
[the remark this answers is not in the document]
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j2/Utils.java:32]: Ignore

File: src/main/java/com/yscope/logging/log4j2/PatternLayoutBufferDestination.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
```
[the remark this answers is not in the document]
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j2/PatternLayoutBufferDestination.java]: Please fix this @p1halani

File: src/main/java/com/yscope/logging/log4j2/AbstractClpIrBufferedRollingFileAppender.java [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+
+    /** @param outputDir The output directory path for log files */
+    private void _setOutputDir(String outputDir) {
+        this.outputDir = outputDir;
+    }
+
+    /** @return The uncompressed size of all log events processed by this appender in bytes. */
+    public long getUncompressedSize() {
+        return uncompressedSizeSinceLastRollover + clpIrFileAppender.getUncompressedSize();
+    }
```
[the remark this answers is not in the document]
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j2/AbstractClpIrBufferedRollingFileAppender.java:136]: Does not apply

File: src/main/java/com/yscope/logging/log4j2/AbstractBufferedRollingFileAppender.java [hunk 4]
```diff
@@ -0,0 +1,869 @@
...
+        );
+        backgroundSyncRunnable.addShutdownRequest();
+        try {
+            backgroundSyncThread.join();
+        } catch (InterruptedException e) {
+            error("Interrupted while joining backgroundSyncThread");
+        }
+
+        super.stop();
+    }
```
[the remark this answers is not in the document]
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j2/AbstractBufferedRollingFileAppender.java]: Please fix @p1halani

File: src/main/java/com/yscope/logging/log4j2/AbstractBufferedRollingFileAppender.java [hunk 5]
```diff
@@ -0,0 +1,869 @@
...
+                boolean deleteFile,
+                Map<String, Object> fileMetadata
+        ) {
+            Request syncRequest = new SyncRequest(
+                    baseName,
+                    logRolloverTimestamp,
+                    deleteFile,
+                    fileMetadata
+            );
+            while (!requests.offer(syncRequest)) {}
```
[the remark this answers is not in the document]
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j2/AbstractBufferedRollingFileAppender.java]: Ignore

File: src/main/java/com/yscope/logging/log4j2/AbstractBufferedRollingFileAppender.java [hunk 6]
```diff
@@ -0,0 +1,869 @@
...
+        private String flushHardTimeoutsInMinutes = null;
+
+        @PluginBuilderAttribute("FlushSoftTimeoutsInSeconds")
+        private String flushSoftTimeoutsInSeconds = null;
+
+        @PluginBuilderAttribute("ShutdownSoftTimeoutInMilliseconds")
+        private long shutdownSoftTimeout = 5000;
+
+        @PluginBuilderAttribute("ShutdownHardTimeoutInSeconds")
+        private long shutdownHardTimeout = 30000;
```
[the remark this answers is not in the document]
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j2/AbstractBufferedRollingFileAppender.java]: Please check if the suggestion is correct. I haven't taken a look at it closely, but I think this is a false positive.

File: src/main/java/com/yscope/logging/log4j2/CompressionPatternLayoutContainer.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                    "Pattern layout must contain timestamp converter"
+            );
+        }
+        String compressionPatternLayoutStr = matcher.group("suffix");
+        compressionPatternLayout = PatternLayout.newBuilder().withPattern(
+                compressionPatternLayoutStr
+        ).build();
+        timestampPattern = matcher.group("timestampPattern");
+        logMsgByteBufferDestination = new PatternLayoutBufferDestination();
+    }
```
[the remark this answers is not in the document]
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j2/CompressionPatternLayoutContainer.java]: This does not apply here.

File: src/test/java/com/yscope/logging/log4j2/ClpIrFileAppenderTest.java [hunk 8]
```diff
@@ -4,9 +4,9 @@
 
 import org.junit.jupiter.api.Test;
 
-public class BoilerplateTests {
+public class ClpIrFileAppenderTest {
     @Test
     public void testGreeting() {
-        assertEquals("Hello, world!", Boilerplate.getGreeting());
+        assertEquals("Hello, world!", "Hello, world!");
     }
```
[the remark this answers is not in the document]
Comment (jackluo923) [reply src/test/java/com/yscope/logging/log4j2/ClpIrFileAppenderTest.java:11]: Does not apply

File: src/main/java/com/yscope/logging/log4j2/PatternLayoutBufferDestination.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    public ByteBuffer drain(final ByteBuffer srcBuf) {
+        // Drain means it ran out of space, so we'll create more space for them
+        ByteBuffer newBuf = ByteBuffer.allocate(buf.capacity() + BUF_ALLOCATION_UNIT);
+        System.arraycopy(buf.array(), 0, newBuf.array(), 0, buf.position());
+        newBuf.position(buf.position());
+
+        // Swap the old buffer with the new buffer
+        buf = newBuf;
+        return buf;
+    }
```
[the remark this answers is not in the document]
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j2/PatternLayoutBufferDestination.java:84]: Not applicable here.

File: src/main/java/com/yscope/logging/log4j2/AbstractBufferedRollingFileAppender.java [hunk 10]
```diff
@@ -0,0 +1,882 @@
...
+                }
+            }
+        }
+
+        /** Adds a shutdown request to the request queue */
+        public void addShutdownRequest() {
+            LOGGER.debug("Adding shutdown request");
+            Request shutdownRequest = new ShutdownRequest();
+            while (!requests.offer(shutdownRequest)) {}
+        }
```
[the remark this answers is not in the document]
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j2/AbstractBufferedRollingFileAppender.java]: Not applicable here.

File: src/main/java/com/yscope/logging/log4j2/AbstractBufferedRollingFileAppender.java [hunk 11]
```diff
@@ -0,0 +1,882 @@
...
+            backgroundFlushThread.start();
+            backgroundSyncThread.start();
+
+            activated = true;
+            super.start();
+        } catch (Exception ex) {
+            error("Failed to activate appender.", ex);
+            super.stop();
+        }
+    }
```
[the remark this answers is not in the document]
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j2/AbstractBufferedRollingFileAppender.java:342]: Not applicable.

File: src/main/java/com/yscope/logging/log4j2/PatternLayoutBufferDestination.java [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+            ByteBuffer newBuf = ByteBuffer.allocate(newBufSize);
+
+            System.arraycopy(buf.array(), 0, newBuf.array(), 0, buf.position());
+            newBuf.position(buf.position());
+
+            buf = newBuf;
+        }
+
+        buf.put(data, offset, length);
+    }
```
[the remark this answers is not in the document]
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j2/PatternLayoutBufferDestination.java:127]: Checks already exist outside of this method in Log4j2 library.

File: src/main/java/com/yscope/logging/log4j2/AbstractClpIrBufferedRollingFileAppender.java [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+                return false;
+            }
+
+            if (outputDir == null || outputDir.isEmpty()) {
+                LOGGER.error("The output directory 'outputDir' must not be null or empty");
+                return false;
+            }
+
+            return true;
+        }
```
[the remark this answers is not in the document]
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j2/AbstractClpIrBufferedRollingFileAppender.java]: Unnecessary

File: src/main/java/com/yscope/logging/log4j2/AbstractClpIrBufferedRollingFileAppender.java [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+
+    /**
+     * Computes a path for the provided log file name
+     *
+     * @param logFileName The log file name
+     * @return The computed log file path
+     */
+    protected String computeLogFilePath(String logFileName) {
+        return Paths.get(outputDir, logFileName).toString();
+    }
```
[the remark this answers is not in the document]
Comment (jackluo923) [reply src/main/java/com/yscope/logging/log4j2/AbstractClpIrBufferedRollingFileAppender.java:216]: Unnecesary

File: src/main/java/com/yscope/logging/log4j2/ClpIrFileAppender.java [hunk 15]
```diff
@@ -0,0 +1,339 @@
...
+ * A Log4j appender that writes log events into a Zstandard-compressed CLP IR stream file.
+ * <p>
+ * Since this appender buffers data in the process of compressing the output, derived appenders
+ * should ensure the appender is closed even when the program exits uncleanly. Otherwise, the
+ * compressed output may be truncated. When the * appender is used directly from Log4j, we install a
+ * shutdown hook for this purpose.
+ */
+public class ClpIrFileAppender extends AbstractAppender implements Flushable {
+    public static final String PLUGIN_NAME = "ClpIrFileAppender";
+    private static final int estimatedFormattedTimestampLength = 0;
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply src/main/java/com/yscope/logging/log4j2/ClpIrFileAppender.java]: @p1halani Is this accurate?

File: src/main/java/com/yscope/logging/log4j2/ClpIrFileAppender.java [hunk 16]
```diff
@@ -0,0 +1,339 @@
...
+                Objects.requireNonNull(clpIrOutputStream).writeLogEvent(
+                        event.getTimeMillis(),
+                        logMsg
+                );
+                uncompressedSizeInBytes += estimatedFormattedTimestampLength + logMsg.limit();
+            } catch (Exception ex) {
+                error("Failed to append log event for appender " + getName(), event, ex);
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply src/main/java/com/yscope/logging/log4j2/ClpIrFileAppender.java]: @p1halani I think comment about line 175 is accurate?

File: src/main/java/com/yscope/logging/log4j2/ClpIrFileAppender.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (p1halani) [reply src/main/java/com/yscope/logging/log4j2/ClpIrFileAppender.java]: It should be 25. Updated

File: src/main/java/com/yscope/logging/log4j2/ClpIrFileAppender.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (p1halani) [reply src/main/java/com/yscope/logging/log4j2/ClpIrFileAppender.java]: Yeah removed.

**Review by kirkrodrigues — APPROVED** [review]

**Pull request merged**